addons/udes_stock/models/stock_picking.py

```diff
@@ -1505,23 +1505,131 @@ def _update_move_lines_and_log_swap(self, move_lines, packs, other_pack):
             specified move lines are a non empty recordset.
         """
         values = {}
-
-        values.update(self._values_for_move_lines_swap(other_pack, values))
+        values.update(self._swap_moveline_info_at_pack_level(other_pack, values))
         move_lines.with_context(bypass_reservation_update=True).write(values)
+        self._swap_moveline_info_below_pack_level(move_lines, other_pack)
         msg_args = (", ".join(packs.mapped("name")), other_pack.name)
         msg = _("Package %s swapped for package %s.") % msg_args
         move_lines.mapped("picking_id").message_post(body=msg)
         _logger.info(msg)
 
-    def _values_for_move_lines_swap(self, other_pack, values):
-        lots = other_pack.mapped("quant_ids.lot_id")
+    def _swap_moveline_info_at_pack_level(self, pack, values):
         values.update({
-            "package_id": other_pack.id,
-            "result_package_id": other_pack.id,
-            "lot_id": lots.id if len(lots) == 1 else False,
+            "package_id": pack.id,
+            "result_package_id": pack.id,
         })
         return values
 
+    def _swap_moveline_info_below_pack_level(self, move_lines, pack):
+        """
+           Swapping of fields which are more granular than pack level - from quant to move_lines
+           specifically, the lots.
+        """
+        MoveLine = self.env["stock.move.line"]
+        sorted_move_lines = move_lines.sorted(key=lambda move_line: move_line.product_uom_qty, reverse=True)
+        move_lines_to_skip = MoveLine.browse()
+        for (product_id, lot_id), quant_ids in pack.mapped("quant_ids").groupby(lambda quant: (quant.product_id, quant.lot_id)):
+            move_lines_to_write = MoveLine.browse()
+            quant_product_quantity = sum(quant_ids.mapped("quantity"))
+            product_move_lines = sorted_move_lines.filtered(
+                lambda move_line: move_line.product_id == product_id and move_line not in move_lines_to_skip
+            )
+            for product_move_line in product_move_lines:
```

Comment:
We can use `product_move_linesmove_lines_for_qty(qty)` instead of this for.

Reply:
done